File: algorithms/algorithms.py

class Algorithms(object):

    def bubble_sort(input):
        if len(input) <= 1:
            return input
        for end in range(len(input)-1,-1, -1):
            for idx in range(0,end):
                if (input[idx+1]) < input[idx]:
                    tmp = input[idx]
                    input[idx] = input[idx+1]
                    input[idx+1] = tmp
        return input

    def merge(self, left, right):
        output = []
        while len(left) > 0 and len(right) > 0:
            if left[0] <= right[0]:
                output.append(left[0])
                left = left[1:]
            else:
                output.append(right[0])
                right = right[1:]

        while len(left) > 0:
            output.append(left[0])
            left = left[1:]

        while len(right) > 0:
            output.append(right[0])
            right = right[1:]
            
        return output


    def merge_sort(self, input):

        if len(input) <= 1:
            return input

        left = []
        right = []

        for idx in range(0, len(input)):
            if idx < len(input) / 2:
                left.append(input[idx])
            else:
                right.append(input[idx])

        left = self.merge_sort(left)
        right = self.merge_sort(right)

        return self.merge(left, right)

File: cpp/src/algorithms.cpp
#include <iostream>
#include <vector>
#include <algorithm>

#include "algorithms.h"

void bubble_sort(std::vector<int> &input)
{
    std::vector<int>::iterator begin = std::begin(input);
    std::vector<int>::iterator end = std::end(input);
    
    for( auto step = end; step != begin; --step) {
        for( auto i = begin; i != step ; ++i) {
            auto &val0 = *i;
            auto &val1 = *(i+1);
            if(val0 > val1) {
                std::swap(val0, val1);
            }
        }
    }
}

void merge_sort(std::vector<int> &input) {

    if (1 >= input.size()) {
        return;
    }

    std::size_t n = input.size()/2;
    std::vector<int> left(input.cbegin(), input.cbegin() + n);
    std::vector<int> right(input.cbegin() + n, input.cend());

    merge_sort(left);
    merge_sort(right);

    std::vector<int> merged = std::vector<int>();
    while ( left.size() > 0 && right.size() > 0 ) {
        auto &lval = *left.begin();
        auto &rval = *right.begin();
        if (lval <= rval) {
            merged.push_back(lval);
            left.erase(left.begin());
        } else {
            merged.push_back(rval);
            right.erase(right.begin());
        }
    }
    
    if (left.size() > 0) {
        std::move(left.begin(), left.end(), std::back_inserter(merged));
    }

    if (right.size() > 0) {
        std::move(right.begin(), right.end(), std::back_inserter(merged));
    }

    input.assign(merged.begin(), merged.end());
}